### ISSUE TITLE
Parallel builds can fail

### ISSUE THREAD

**Issue by M4rkD** [RcppCore/Rcpp#1079]
Starting multiple copies of an Rscript which uses Rcpp in parallel causes issues on linux systems (not tested on Windows). After some investigatin is seems that this is happening in this function of Attributes.R (and also attributes.cpp):
```
.sourceCppDynlibWriteCache <- function(cacheDir, cache) {
    index_file <- file.path(cacheDir, "cache.rds")
    save(cache, file = index_file, compress = FALSE)
}
```
which is called from
```
.sourceCppDynlibInsert <- function(cacheDir, file, code, dynlib) {
    cache <- .sourceCppDynlibReadCache(cacheDir)
    index <- .sourceCppFindCacheEntryIndex(cache, file, code)
    if (is.null(index))
        index <- length(cache) + 1
    cache[[index]] <- list(file = file, code = code, dynlib = dynlib)
    .sourceCppDynlibWriteCache(cacheDir, cache)
}
```
Looking at the system calls, it seems to me that the update of the cache isn't being atomically (the update is done under the hood as multiple smaller writes). I wonder if it's the case that the save function itself isn't suitable for parallel use?

Which could cause corrupted "cache.rds" files in parallel, which may be the case.

It seems to me that reading cache, then acted on it, then written non-atomically mean that Rcpp can't support parallel execution of the same script reliably. In fact, we see a number of crashes, in which this file is the only file being accesses. This happens both with and without previous compilation.
 
In addition, it seems to me that the cache.rds file is written even when there is no build required (see the sourceCpp function of attributes.cpp). So even building beforehand wouldn't allow reliable parallel execution.

One could imagine solutions based on lock files, or on splitting the cache across multiple files and trying to use atomic file operations. But, I think by far the easiest (and most robust) solution would be for each process to somehow have a uniquely named cache.rds file.

**Comment by eddelbuettel** [comment]
Hm, maybe worth keeping in mind that _Writing R Extensions_ is, as I recall, not exactly encouraging parallelism _within_ a package build, but that parallel installation or build of packages is possible (and well supported via option `Ncpus`.

**Comment by kevinushey** [comment]
Some thoughts:

- The cache directory used by Rcpp can be controlled either explicitly via parameter, or with an option. If controlling it is necessary, you could do so manually yourself. See the documentation in `?sourceCpp`.

- By default, each R session gets its own temporary directory, and so each process should be using an independent cache. Are you not seeing this? Or are you doing something like fork-based parallelism?

Let me know if I'm misunderstanding something in your use case.

(In addition, a minimally reproducible example would be useful in helping us further understand + diagnose the issue)

**Comment by M4rkD** [comment]
Thank for the quick replies.

I knew about the ability to set the cache directory (this is in fact the workaround that we're using). I wasn't aware that this cache was newly created in each session. The author of the code we're running sets a constant value for cacheDir by default, so I didn't realise it defaults to a new build per session.

This approach seems convenient to me when performance matters on large numbers of runs, since builds can be expensive.

I think the ideal "nice to have" from this point of view would be the ability to share the builds products, to avoid recompilation, but still allow multiple processes to run at once. Which I think probably means not sharing the "cache.rds" file.

For example a way to configure the filename for "cache.rds" independently of the directory would allow this.

**Comment by eddelbuettel** [comment]
I suggest we close this for being out of scope.

Rcpp does not "invent" or "modify" the build system. `R CMD <verb_here>` it is for different verbs.  If and when you desire changes in behavior with these, you should take it to R Core.

But please feel free to reopen if you anything actionable we can influence.
